### ISSUE TITLE
MemoryStream is not [Serializable]

### ISSUE THREAD

**Issue by dasetser** [dotnet/coreclr#17460]
MemoryStream is not marked with the SerializableAttribute and fails to be deserialized. Without this attribute we're unable to create a WCF client to communicate with existing services that return MemoryStreams.

This was reported in  dotnet/wcf#2307.

**Comment by morganbr** [comment]
Do you have a sense of why somebody would want to serialize a Stream? It doesn't actually stream anything if they do that, so serializing a MemoryStream is mostly just an expensive way to pass a byte array.

**Comment by dasetser** [comment]
I agree it's just an expensive way to pass a byte array. On the original issue I suggested if they own the WCF service using a MemoryStream they should change it to use a Stream to use a Streamed transfer mode instead, which would make a lot more sense.

The problem is if someone needs to create a client to communicate with a service they don't own they don't have the choice to fix the issue and do something that makes more sense, and have to use a workaround to get it to deserialize at all.

**Comment by RussKeldorph** [comment]
@danmosemsft

**Comment by jkotas** [comment]
I do not think we want to start making Streams serializable for this corner case.

The workaround suggested in https://github.com/dotnet/wcf/issues/2307#issuecomment-379359370 seems reasonable to me.

**Comment by mconnew** [comment]
@jkotas, this is a dangerous behavior difference between .Net Core and the full framework because the failure is silent. If you have a MemoryStream as part of data which you wish to serialize, you might legitimately have a case where the deserialized MemoryStream contains no data. There's no way for code to know whether the MemoryStream being empty is legitimate or not without directly examining the original XML which existing code wouldn't do because why would you? Serialization just worked on the full framework. This difference between frameworks can cause a silent loss of data which is a very bad situation for a framework to be in. Even worse, if a common library runs on netstandard and netcore, the same code will have different behavior.  
I am strongly against the idea of removing an attribute for philosophical reasons when that removal has very real data loss consequences.

**Comment by danmosemsft** [comment]
@mconnew I have this question also:
> Is there a good reason why is your serializer dropping data silently to the floor? Would it be possible to add a better diagnostic for this? The same data loss can occur for any of the many hundred types that are not binary serializable in .NET Core, not just MemoryStream.

**Comment by mconnew** [comment]
@danmosemsft, the short answer is versioning. The long answer can be found in [this article](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/data-contract-versioning). But basically this is an intentional design to make forwards compatibility painless. When versioning forwards the presumption is you can't modify existing code with the previous version (you might not own those clients for example) so it's the responsibility of the dev to version forward in a way which doesn't break existing clients. If a new field being added breaks existing code without modification, then versioning would be a lot more painful, and in some cases impossible.

**Comment by danmosemsft** [comment]
@mconnew MemoryStream aside, there is probably over 100 types that are no longer marked [Serializable] so this is a broader issue.

I understand that you want code to robustly handle unexpected extra fields. I don't see mention however in the article of the specific case where the deserializer recognizes the type but it's own implementation is not [Serializable] (while evidently the serializer's copy was). Could that represent anything other than a breaking change in serializability and as such is that something that could be specially flagged? Ultimately it is a policy decision whether it works by default.

**Comment by mconnew** [comment]
The Serializable attribute to DCS simply means "it's safe to serialize my fields whether they are private or not". Without it, DCS only serializes properties. The xml representation doesn't imply what the original class looked like, it's not meant to be a direct equivalent of the class like BinaryFormatter produces. I'll give you a simple example with two versions of the same class, one with the Serializable attribute and one without and they will serialize identically with DCS.  
With attribute:
```csharp
[Serializable]
public class Foo
{
    public string Cat;
    public string Dog;
}
```
Without attribute:
```csharp
public class Foo
{
    private string _cat;
    private string _dog;

    public string Cat
    {
        get { return _cat; }
        set { _cat = value; }
    }

    public string Dog
    {
        get { return _dog; }
        set { _dog = value; }
    }
}
```
There is no way to detect that the source had the Serializable attribute and locally you don't. They could have even come from completely different types. In fact, that's how the workaround in the WCF issue works. It's a completely different type configured to serialize the same way as MemoryStream. If there were a way to differentiate, then throwing when that's the case would be a major breaking change as you might have someone who coded their class like the first example above and changed it to the second. Detecting that you are now dealing with a class not marked as Serializable would break someone trying to eliminate the Serializable attribute in their code. The only way to detect would be by convention. So which convention do you detect and throw on? When elements exist with an underscore prefix (e.g. `_length`), or elements with m_ prefix (e.g. `m_length`)? What about with a convention of a type prefix (e.g. `strName`) or lowercase first letter? Also the serializer doesn't recognize the type (it's not like BinaryFormatter), you construct a serializer for the specific type and the generated serializer can deserialize xml which matches the class and what it's attributes specify.  
You are right about it being a policy decision whether to throw if there are fields you don't know about. You can argue the merits either way but the reality is it was previously chosen that the policy is that versioning should be easy and not require going back and changing existing code. To change the policy now would mean breaking a very large number of users of DCS. It would break anyone who is using the versioning capabilities and they would need to change their code to tell it to work in a versioning compatible way.  
@jkotas, I would like to restate the deep technical decision that was stated in the linked bug. Stephen said:
> But it comes with some serious implications, and it's becoming clear we took things too far. Doing this makes it very easy for code to take dependencies on private implementation details and will end up binding our hands for improvements/optimizations/refactorings/etc. in the future.

Another way of saying this is that making types backwards and forwards compatible is difficult and we don't want to do it so we're going to forcibly break compatibility between .Net Core and the full framework so we don't need to be concerned about backwards and forwards compatibility in the future.

I would categorize this as a philosophical decision. I would be very surprised if there are many classes with the Serializable attribute where you wouldn't be able to use the serialization attributes and interfaces to make internal changes still be compatible (and it would make sense to serialize the type to begin with).  
Were any of the other serializers in the framework considered when making this decision? I only saw discussion of BinaryFormatter in the linked bug but this decision has repercussions beyond BinaryFormatter. The link in the description where it lists the types which can be serialized is an article about BinaryFormatter so it seems the concern was only for BinaryFormatter and other serializers were forgotten about in that discussion. **If implications to other serializers weren't considered in that discussion, then I believe the decision needs to be revisited as there are important consequences which weren't discussed.**
